File: Intatis-Android/shared/src/main/kotlin/com/intatis/shared/ConfigStore.kt
package com.intatis.shared

import java.io.File
import java.nio.file.Path
import java.nio.file.Paths
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

object ConfigStore {
    private const val ConfigDir = ".config"
    private const val AppDir = "intatis"
    private const val ConfigFile = "config.json"
    private val json = Json {
        ignoreUnknownKeys = true
    }

    @Volatile
    private var configHomeOverride: File? = null

    fun configureConfigDirectory(directory: File) {
        configHomeOverride = directory
    }

    fun clearConfigDirectoryOverride() {
        configHomeOverride = null
    }

    val configFolder: String
        get() {
            val override = configHomeOverride
            if (override != null) return override.absolutePath

            val home = System.getProperty("user.home")
                ?: System.getProperty("user.dir")
                ?: "."
            return File(home, "$ConfigDir/$AppDir").absolutePath
        }

    val configPath: String
        get() = Paths.get(configFolder, ConfigFile).toString()

    fun load(): IntatisConfig {
        val fileValues = loadFileValues()

        fun value(envKey: String, fileKey: String, fallback: String): String {
            val env = System.getenv(envKey)
            if (!env.isNullOrBlank()) return env
            val fromFile = fileValues[fileKey]
            if (!fromFile.isNullOrBlank()) return fromFile
            return fallback
        }

        val baseUrl = value("INTATIS_BASE_URL", "baseUrl", "https://api.openai.com/v1")
        val apiKey = value("INTATIS_API_KEY", "apiKey", "")
        val model = value("INTATIS_MODEL", "model", "gpt-4o-mini")
        val reasoning = System.getenv("INTATIS_REASONING") ?: fileValues["reasoning"]

        val modeString = value("INTATIS_MODE", "defaultMode", IntatisMode.Chat.name).lowercase()
        val mode = try {
            IntatisMode.valueOf(modeString.replaceFirstChar { it.uppercaseChar() })
        } catch (_: IllegalArgumentException) {
            IntatisMode.Chat
        }

        val workspace = System.getenv("INTATIS_WORKSPACE") ?: fileValues["workspace"]
        val usageValue = value("INTATIS_USAGE", "includeUsage", "1").lowercase()
        val includeUsage = !(usageValue == "0" || usageValue == "false" || usageValue == "off")

        return IntatisConfig(
            baseUrl = baseUrl,
            apiKey = apiKey,
            model = model,
            reasoning = reasoning?.trim()?.takeIf { it.isNotEmpty() && it != "off" },
            defaultMode = mode,
            workspace = workspace?.takeIf { it.isNotBlank() },
            includeUsage = includeUsage,
        )
    }

    fun save(config: IntatisConfig) {
        val folder = File(configFolder)
        folder.mkdirs()

        val payload = buildJsonObject {
            put("baseUrl", JsonPrimitive(config.baseUrl))
            put("apiKey", JsonPrimitive(config.apiKey))
            put("model", JsonPrimitive(config.model))
            put("defaultMode", JsonPrimitive(config.defaultMode.name))
            config.reasoning?.let { put("reasoning", JsonPrimitive(it)) }
            config.workspace?.takeIf { it.isNotBlank() }?.let { put("workspace", JsonPrimitive(it)) }
            put("includeUsage", JsonPrimitive(config.includeUsage))
        }

        val target = File(configPath)
        target.parentFile.mkdirs()
        target.writeText(payload.toString())

        try {
            target.setReadable(false)
            target.setReadable(true, true)
            target.setWritable(true, true)
        } catch (_: Throwable) {
            // best effort
        }
    }

    private fun loadFileValues(): Map<String, String> {
        val file = File(configPath)
        if (!file.exists()) return emptyMap()

        return try {
            val element = json.parseToJsonElement(file.readText()).jsonObject
            element.entries.associate { entry ->
                entry.key to valueToString(entry.value)
            }
        } catch (_: Throwable) {
            emptyMap()
        }
    }

    private fun valueToString(element: kotlinx.serialization.json.JsonElement): String {
        return when (element) {
            is JsonObject -> json.encodeToString(JsonObject.serializer(), element)
            else -> element.jsonPrimitive.contentOrNull?.trim() ?: element.toString().trim('"')
        }
    }
}

File: Intatis-Windows/src/Intatis.Shared/ConfigStore.cs
using System.Text.Json;

namespace Intatis.Windows.Shared;

public static class ConfigStore
{
    private const string AppFolder = "Intatis";
    private const string AppName = "Intatis-Windows";
    private const string ConfigFileName = "config.json";

    public static string ConfigFolder =>
        Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), AppFolder, AppName);

    public static string ConfigPath =>
        Path.Combine(ConfigFolder, ConfigFileName);

    public static IntatisConfig Load()
    {
        var fileValues = LoadFromFile();

        string value(string envKey, string fileKey, string? fallback)
        {
            var env = Environment.GetEnvironmentVariable(envKey);
            if (!string.IsNullOrWhiteSpace(env))
                return env;
            if (fileValues.TryGetValue(fileKey, out var v) && !string.IsNullOrWhiteSpace(v))
                return v;
            return fallback ?? string.Empty;
        }

        var baseUrl = value("INTATIS_BASE_URL", nameof(BaseUrl), "https://api.openai.com/v1");
        var apiKey = value("INTATIS_API_KEY", nameof(ApiKey), string.Empty);
        var model = value("INTATIS_MODEL", nameof(Model), "gpt-4o-mini");
        var reasoning = value("INTATIS_REASONING", nameof(Reasoning), null);
        var modeValue = value("INTATIS_MODE", nameof(DefaultMode), IntatisMode.Chat.ToString());
        var workspace = value("INTATIS_WORKSPACE", nameof(Workspace), string.Empty);
        var usageValue = value("INTATIS_USAGE", nameof(IncludeUsage), "1").ToLowerInvariant();
        var includeUsage = usageValue is not ("0" or "false" or "off");

        var mode = Enum.TryParse(modeValue, true, out IntatisMode parsedMode)
            ? parsedMode
            : IntatisMode.Chat;

        return new IntatisConfig(
            baseUrl,
            apiKey,
            model,
            string.IsNullOrWhiteSpace(reasoning) ? null : reasoning,
            mode,
            string.IsNullOrWhiteSpace(workspace) ? null : workspace,
            includeUsage);
    }

    public static void Save(IntatisConfig config)
    {
        var payload = new Dictionary<string, string>
        {
            [nameof(BaseUrl)] = config.BaseUrl,
            [nameof(ApiKey)] = config.ApiKey,
            [nameof(Model)] = config.Model,
            [nameof(DefaultMode)] = config.DefaultMode.ToString().ToLowerInvariant()
        };

        if (!string.IsNullOrWhiteSpace(config.Reasoning))
            payload[nameof(Reasoning)] = config.Reasoning!;
        if (!string.IsNullOrWhiteSpace(config.Workspace))
            payload[nameof(Workspace)] = config.Workspace!;
        payload[nameof(IncludeUsage)] = config.IncludeUsage ? "1" : "0";

        Directory.CreateDirectory(ConfigFolder);
        var json = JsonSerializer.Serialize(payload, new JsonSerializerOptions { WriteIndented = true });
        File.WriteAllText(ConfigPath, json);
        try
        {
            File.SetAttributes(ConfigPath, FileAttributes.Hidden | File.GetAttributes(ConfigPath));
        }
        catch
        {
            // Best effort only. On some systems attributes may be restricted.
        }
    }

    private static Dictionary<string, string> LoadFromFile()
    {
        if (!File.Exists(ConfigPath))
            return new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);

        try
        {
            var raw = File.ReadAllText(ConfigPath);
            return JsonSerializer.Deserialize<Dictionary<string, string>>(raw, new JsonSerializerOptions())
                   ?? new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        }
        catch
        {
            return new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        }
    }
}

